include/seqan3/range/decorator/gap_decorator_anchor_set.hpp

```diff
@@ -0,0 +1,767 @@
+                left_gap_end = (*anchor_set_it).first + (*anchor_set_it).second -
+                               ((anchor_set_it != (*host).anchors.begin()) ?
+                                    (*std::prev(anchor_set_it)).second : 0);
+                ++anchor_set_it;
+            }
+
+            return *this;
+        }
+
+        //!\brief Pre-decrement, returns updated iterator.
+        constexpr gap_decorator_anchor_set_iterator & operator--() noexcept
+        {
+            assert((host)); // host is set
+            --pos;
+
+            if (pos < left_gap_end)
+            {   // there was no gap before but we arrive at the left gap and have to update the variables.
+                if (anchor_set_it != (*host).anchors.begin())
+                {
+                    --anchor_set_it;
+                    if (anchor_set_it != (*host).anchors.begin())
+                    {
+                        auto prev = std::prev(anchor_set_it);
+                        left_gap_end = (*prev).first + (*prev).second -
+                                       ((prev != (*host).anchors.begin()) ? (*std::prev(prev)).second : 0);
+                    }
+                    else // [[unlikely]]
+                    {
+                        left_gap_end = 0;
+                    }
+                }
+            }
+            else if (anchor_set_it == (*host).anchors.end() || pos < (*anchor_set_it).first)
+            {   // we are neither at the left nor right gap
+                --ungapped_view_pos;
+            }
+            // else -> no op (we are still within the right gap stretch)
+
+            return *this;
+        }
+
+        //!\brief Post-increment, returns previous iterator state (delegates to pre-increment).
+        constexpr gap_decorator_anchor_set_iterator operator++(int) noexcept
+        {
+            gap_decorator_anchor_set_iterator cpy{*this};
+            ++(*this);
+            return cpy;
+        }
+
+        //!\brief Post-decrement, returns previous iterator state (delegates to pre-decrement).
+        constexpr gap_decorator_anchor_set_iterator operator--(int) noexcept
+        {
+            gap_decorator_anchor_set_iterator cpy{*this};
+            --(*this);
+            return cpy;
+        }
+        //!\}
+
+        /*!\name Reference/Dereference operators
+         * \{
+        */
+        //!\brief Dereference operator returns element currently pointed at.
+        constexpr reference operator*() const noexcept
+        {
+            if (pos >= left_gap_end && (anchor_set_it == (*host).anchors.end() || pos < (*anchor_set_it).first))
+                return value_type{(*host).ungapped_view[ungapped_view_pos]};
+            else
+                return gap{};
+        }
+
+        //!\brief Returns pointer to this iterator.
+        constexpr pointer operator->() const noexcept
+        {
+            // ?
```

Comment:
Wow, this took me longer to research than I thought. 

This question sums up your use-case:
https://stackoverflow.com/questions/27582845/how-to-implement-operator-for-an-iterator-that-constructs-its-values-on-demand

In short: Strictly speaking you cannot make this iterator model anything stronger than InputIterator **according to the old language specification**. AFAICT the new requirements (post ranges) concepts make no assumptions about `operator->` at all.

There are two ways to deal with this:
1. just remove the operator and also the `::pointer` member type
2. store a member a variable of type `value_type` in the iterator and have every increment/decrement store the current value in that; have `operator->` return the address of that. 

The benefit of solution 2 is that old algorithms would still consider this an iterator and also you can just return that value on `operator*` and don't need the check.

You could also ask on stack overflow of course.